test/refactoredTestSetup.ts

```diff
@@ -0,0 +1,336 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import fs from 'fs';
+import path from 'path';
+import * as utilities from './utilities/index.ts';
+import { EnvironmentSettings as Env } from './environmentSettings.ts';
+import { ProjectConfig, ProjectShapeOption } from './utilities/index.ts';
+import { fileURLToPath } from 'url';
+
+const __filename = fileURLToPath(import.meta.url);
+const __dirname = path.dirname(__filename);
+
+export class refactoredTestSetup {
+  public testSuiteSuffixName: string = '';
+  public tempFolderPath = path.join(__dirname, '..', 'e2e-temp');
+  public projectFolderPath: string | undefined;
+  private aliasAndUserNameWereVerified = false;
+  public scratchOrgAliasName: string | undefined;
+  public scratchOrgId: string | undefined;
+
+  public constructor() {
```

Comment:
What do you think about passing test config in the constructor and then referencing it in the other functions? 

My thinking would be to limit the possibility of changing the test config using instance methods.

With the current impl I can do things like
```
const ts = new refactoredTestSetup();
await ts.setup(config1);
await ts.setup(config2);
await ts.setUpTestingWorkspace(projectConfig1);
await ts.setUpTestingWorkspace(projectConfig2);
```
While nonsensical, it does show that the test setup is not controlling the config of the test.

This a balance between flexibility and config control, so think about finding a balance.

This also could be done with the factory pattern.

```
public class TestSetup {
  private constructor(public readonly testReqConfig: utilities.TestReqConfig) {}
  public static async setup(testReqConfig: utilities.TestReqConfig): Promise<void> {
    // do setup stuff
  }
  // instance methods that we allow on the test setup
  public async teardown() {}
}
```
then
```
const ts = await TestSetup.setup(testConfig);
// do test related stuff and mutating the test project context is not allowed via `ts`
```

Our discussions thus far have been around one test/one project. How do we and why would we handle changing the workspace to an entirely different project?

Reply:
An example of calling `setUp` multiple times within one test is `pushAndPull.e2e.ts` where we create a second project to connect to the same scratch org and pull the source. So with this case as an example, I think testSetup should be associated with one test rather than one project setting.

The reason why I do not use constructor is mostly that we cannot call async functions within it. 

Does this make sense to you?